File: rtl/NSArray.cpp
#include "NSArray.h"
#include "NSString.h"
#include "context.h"
#include <algorithm>
#include <iostream>

// Статический ClassInfo
ClassInfo* NSArray::s_classInfo = nullptr;

// ------------------- Конструкторы -------------------
NSArray::NSArray() : array() { }

NSArray::NSArray(const NSArray& other) : array(other.array) { }

NSArray::NSArray(const std::vector<std::shared_ptr<NSObject>>& objects) : array(objects) { }

// ------------------- Статические методы -------------------
std::shared_ptr<NSArray> NSArray::arrayStatic() {
    return std::make_shared<NSArray>();
}

std::shared_ptr<NSArray> NSArray::arrayWithArrayStatic(const NSArray& array) {
    return std::make_shared<NSArray>(array);
}

std::shared_ptr<NSArray> NSArray::arrayWithObjectsStatic(const std::vector<std::shared_ptr<NSObject>>& objects) {
    return std::make_shared<NSArray>(objects);
}

std::shared_ptr<NSArray> NSArray::arrayWithObjectStatic(std::shared_ptr<NSObject> object) {
    return std::make_shared<NSArray>(std::vector<std::shared_ptr<NSObject>>{object});
}

// ------------------- Динамические методы -------------------
std::shared_ptr<NSArray> NSArray::arrayByAddingObjectDynamic(std::shared_ptr<NSObject> object) {
    auto res = array;
    res.push_back(object);
    return std::make_shared<NSArray>(res);
}

std::shared_ptr<NSArray> NSArray::arrayByAddingObjectsFromArrayDynamic(const NSArray& anotherArray) {
    auto res = array;
    res.insert(res.end(), anotherArray.array.begin(), anotherArray.array.end());
    return std::make_shared<NSArray>(res);
}

std::shared_ptr<NSObject> NSArray::objectAtIndexDynamic(int index) {
    if (index < 0 || index >= static_cast<int>(array.size())) return nullptr;
    return array[index];
}

int NSArray::countDynamic() const {
    return static_cast<int>(array.size());
}

int NSArray::containsObjectDynamic(std::shared_ptr<NSObject> object) {
    return std::any_of(array.begin(), array.end(),
                       [&](auto& o) { return o == object; }) ? 1 : 0;
}

std::shared_ptr<NSString> NSArray::componentsJoinedByStringDynamic(const NSString& separator) {
    std::string sep = separator.cStringDynamic().get();
    std::string res;
    for (size_t i = 0; i < array.size(); ++i) {
        auto desc = array[i]->descriptionDynamic();
        res += desc;
        if (i != array.size() - 1) res += sep;
    }
    return NSString::stringWithCStringStatic(res.c_str());
}

std::shared_ptr<NSObject> NSArray::firstObjectDynamic() {
    return array.empty() ? nullptr : array[0];
}

std::shared_ptr<NSObject> NSArray::lastObjectDynamic() {
    return array.empty() ? nullptr : array.back();
}

std::shared_ptr<NSObject> NSArray::firstObjectCommonWithArrayDynamic(const NSArray& otherArray) {
    for (auto& o : otherArray.array) {
        if (containsObjectDynamic(o)) return o;
    }
    return nullptr;
}

int NSArray::indexOfObjectDynamic(std::shared_ptr<NSObject> object) {
    for (size_t i = 0; i < array.size(); ++i) {
        if (array[i] == object) return static_cast<int>(i);
    }
    return -1; // NSNotFound
}

int NSArray::isEqualToArrayDynamic(const NSArray& otherArray) {
    if (array.size() != otherArray.array.size()) return 0;
    for (size_t i = 0; i < array.size(); ++i) {
        if (array[i]->isEqualDynamic(otherArray.array[i].get()) == 0) return 0;
    }
    return 1;
}

NSArray* NSArray::initDynamic() {
    NSObject::initDynamic();
    return this;
}

std::string NSArray::descriptionDynamic() {
    return "NSArray instance";
}

// ------------------- Инициализация ClassInfo -------------------
void initNSArrayClassInfo(ClassInfo* info) {
    NSArray::s_classInfo = info;
}

File: rtl/NSArray.java
package rtl;

import java.util.ArrayList;
import java.util.List;

public class NSArray extends NSObject {

    private List<NSObject> array;

    // ------------------- Конструкторы -------------------
    public NSArray() {
        this.array = new ArrayList<>();
    }

    public NSArray(NSArray other) {
        this.array = new ArrayList<>(other.array);
    }

    public NSArray(List<NSObject> objects) {
        this.array = new ArrayList<>(objects);
    }

    // ------------------- Статические методы -------------------
    public static NSArray arrayStatic() {
        return new NSArray();
    }

    public static NSArray arrayWithArrayStatic(NSArray array) {
        return new NSArray(array);
    }

    public static NSArray arrayWithObjectsStatic(NSObject... objects) {
        NSArray nsArray = new NSArray();
        for (NSObject obj : objects) {
            nsArray.array.add(obj);
        }
        return nsArray;
    }

    public static NSArray arrayWithObjectStatic(NSObject object) {
        return arrayWithObjectsStatic(object);
    }

    // ------------------- Динамические методы -------------------
    public NSArray arrayByAddingObjectDynamic(NSObject object) {
        NSArray res = new NSArray(this);
        res.array.add(object);
        return res;
    }

    public NSArray arrayByAddingObjectsFromArrayDynamic(NSArray anotherArray) {
        NSArray res = new NSArray(this);
        res.array.addAll(anotherArray.array);
        return res;
    }

    public NSObject objectAtIndexDynamic(int index) {
        if (index < 0 || index >= array.size()) return null;
        return array.get(index);
    }

    public int countDynamic() {
        return array.size();
    }

    public int containsObjectDynamic(NSObject object) {
        return array.contains(object) ? 1 : 0;
    }

    public NSString componentsJoinedByStringDynamic(NSString separator) {
        StringBuilder sb = new StringBuilder();
        String sep = separator.toString();
        for (int i = 0; i < array.size(); i++) {
            sb.append(array.get(i).descriptionDynamic());
            if (i != array.size() - 1) sb.append(sep);
        }
        return NSString.stringWithStringStatic(new NSString(sb.toString()));
    }

    public NSObject firstObjectDynamic() {
        return array.isEmpty() ? null : array.get(0);
    }

    public NSObject lastObjectDynamic() {
        return array.isEmpty() ? null : array.get(array.size() - 1);
    }

    public NSObject firstObjectCommonWithArrayDynamic(NSArray otherArray) {
        for (NSObject o : otherArray.array) {
            if (array.contains(o)) return o;
        }
        return null;
    }

    public int indexOfObjectDynamic(NSObject object) {
        int index = array.indexOf(object);
        return index >= 0 ? index : -1; // NSNotFound
    }

    public boolean isEqualToArrayDynamic(NSArray otherArray) {
        if (array.size() != otherArray.array.size()) return false;
        for (int i = 0; i < array.size(); i++) {
            if (!array.get(i).isEqualDynamic(otherArray.array.get(i))) return false;
        }
        return true;
    }


    @Override
    public NSArray initDynamic() {
        super.initDynamic();
        return this;
    }

    @Override
    public String descriptionDynamic() {
        return "NSArray instance";
    }

    // ------------------- Статический ClassInfo -------------------
    public static ClassInfo s_classInfo;

    public static void initClassInfo(ClassInfo info) {
        s_classInfo = info;
    }
}
